common/os_calls.c

```diff
@@ -1455,26 +1456,24 @@ g_sck_socket_ok(int sck)
 int
 g_sck_can_send(int sck, int millis)
 {
-    fd_set wfds;
-    struct timeval time;
-    int rv;
-
-    time.tv_sec = millis / 1000;
-    time.tv_usec = (millis * 1000) % 1000000;
-    FD_ZERO(&wfds);
-
+    int rv = 0;
     if (sck > 0)
     {
-        FD_SET(((unsigned int)sck), &wfds);
-        rv = select(sck + 1, 0, &wfds, 0, &time);
+        struct pollfd pollfd;
 
-        if (rv > 0)
+        pollfd.fd = sck;
+        pollfd.events = POLLOUT;
+        pollfd.revents = 0;
+        if (poll(&pollfd, 1, millis) > 0)
```

Comment:
You are missing a case here: When `poll()` returns `-1` and errno is set to `EINTR`, then you need to call `poll()` again, i.e. do-while loop it (applies to other poll calls in the other lines as well).

Reply:
Thanks @pnowack 

I can't see how this change differs from the existing `select()` code - am I missing something? In the original code the select will return -1/EINTR as will poll in the new code. The result of both calls will be to return 0 early.

The calls with a zero timeout will not return EINTR(?) so they won't need a loop.

That leaves the following according to my analysis:-

|Call                    | Analysis |
|----------------------|-------------|
|os_calls.c:1494  | Same as os_calls.c:1467 (i.e. behaviour unchanged) |
|os_calls.c:1927  | Original code also returns early on EINTR |
|xrdpapi.c:491  | Behaviour unchanged |
|xrdpapi.c:512 | Behaviour unchanged |
|xrdp-ssh-agent.c:294 | Behaviour unchanged |
|xrdp-ssh-agent.c:400 | Behaviour unchanged |
|xrdp_pcsc.c:277 | Behaviour unchanged |
|xcon.c:55 | Behaviour unchanged |
|gtcp.c:278 | Behaviour unchanged |

I've also examined what the effect of the existing behaviour would be:-

|Call                    | Behaviour on EINTR |
|----------------------|-------------|
|os_calls.c:1467  | Wait finishes early. Socket is not read until next poll |
|os_calls.c:1494  | Wait finishes early. Socket is not read until next poll |
|os_calls.c:1927  | Wait finishes early and sockets are re-polled |
|xrdpapi.c:491  | **Program can fail** |
|xrdpapi.c:512 | **Program can fail** |
|xrdp-ssh-agent.c:294 | No effect |
|xrdp-ssh-agent.c:400 | No effect |
|xrdp_pcsc.c:277 | No effect, because of unusual program logic in following poll() |
|xcon.c:55 | **Program fails** |
|gtcp.c:278 | Behaviour unchanged |

So there's a case here for fixing xrdpapi.c and xcon.c. I can certainly do that.

There's also the small problem that on an interrupt, we can't easily work out how much of the timeout has expired. A retry loop will result in the timeout possibly being exceeded, or in the pathological case, never completing.

Please come back to me if I'm missing something or I'm not being clear.